File: Stage24/Python/Problem1780.py
import sys
input = sys.stdin.readline

minus = 0
zero = 0
plus = 0

def countPaper(x, y, size):
    global minus, zero, plus
    
    first = paper[x][y]
    same = True
    
    for i in range(x, x + size):
        for j in range(y, y + size):
            if paper[i][j] != first:
                same = False
                break
        if not same:
            break
    
    if same:
        if first == -1:
            minus += 1
        elif first == 0:
            zero += 1
        else:
            plus += 1
        return
    
    newSize = size // 3
    for i in range(3):
        for j in range(3):
            countPaper(x + i * newSize, y + j * newSize, newSize)

def main():
    global paper
    
    N = int(input())
    paper = [list(map(int, input().split())) for _ in range(N)]
    
    countPaper(0, 0, N)
    
    print(minus)
    print(zero)
    print(plus)
    
if __name__ == "__main__":
    main()

File: Stage24/C/Problem1780.c
#include <stdio.h>

#define MAX 2200

int N;
int paper[MAX][MAX];
int minus = 0, zero = 0, plus = 0;

void countPaper(int x, int y, int size) {
    int first = paper[x][y];
    int same = 1;

    for (int i = x; i < x + size; i++) {
        for (int j = y; j < y + size; j++) {
            if (paper[i][j] != first) {
                same = 0;
                break;
            }
        }
        if (!same) break;
    }

    if (same) {
        if (first == -1) minus++;
        else if (first == 0) zero++;
        else plus++;
        return;
    }

    int newSize = size / 3;
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            countPaper(x + i * newSize, y + j * newSize, newSize);
        }
    }
}

int main(void) {
    scanf("%d", &N);

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            scanf("%d", &paper[i][j]);
        }
    }

    countPaper(0, 0, N);

    printf("%d\n%d\n%d\n", minus, zero, plus);
    return 0;
}